### ISSUE TITLE
License scanner

### ISSUE THREAD

**Issue by itaysk** [aquasecurity/trivy#2303]
By popular request we want Trivy to scan licenses in the scan target. This is not strictly a security scan, but it does relate to the overall risk posture of the application and also something that's commonly done as part of SCA (similar to SBOM which we already do).

Today Trivy already relates to licenses in a limited way (mostly to satisfy SBOM formats): only in container scanning, and only for some languages (where dependencies source code is available).

In line with the way other Trivy scanner work, this should show "problems" to be addressed, and not just facts about the scanned artifact. For example, vulnerability scanner doesn't show all packages by default, it shows only vulnerable packages. Similarly, license scanner doesn't show all discovered licenses by default, only potentially risky licenses (see below).

Another consideration is discovering licenses in packages vs in the filesystem. Both are interesting and serve different use-cases:
1. As a developer, I would like to verify that the software dependencies I used are legally safe in my use case. If there is an issue, I'd like to see which package was the culprit. For example, I have imported a AGPL licensed package which conflicts with the license of my application. I should either address the license requirements or remove the dependency from my app.
2. As a system operator, I would like to ensure that an artifact that I'm about to run/install is legally safe to use. If there's an issue, I'd like to be know that I should not run it, and possibly some information that can be communicated to the relevant maintainer. For example, I'm going to run some container from DockerHub, but apparently it contains a component that cannot be distributed as such. I should be alerted not to use this container, and perhaps contact the developer with technical information to seek advice.

Google had created useful framework that we can follow: https://opensource.google/documentation/reference/thirdparty/licenses
They classify third party dependencies into categories: restricted, restricted_if_statically_linked, reciprocal, notice, permissive.
They also open sourced the library used to discover licenses which also have a list of "forbidden" licenses, I presume for commercial use: https://github.com/google/licenseclassifier/blob/842c0d70d7027215932deb13801890992c9ba364/license_type.go
so I would create this ladder:

1. non-standard	
2. non-commercial (Google's forbidden)	
3. restricted	
4. restricted_if_statically_linked
5. reciprocal	
6. notice
7. permissive
8. unencumbered

And suggest to put the "risky license" threshold by default at 4 (meaning 1-4 including are bad)

to summarize:

1. Create a new scanner (`--security-check`) called `license`
2. Table for usecase#1: 
    a. License: normalized license name (in SPDX License ID format)
    b. License information: License Classification, link to `https://spdx.org/licenses/${LICENSEID}.html`
    c. Package: path to package within package manager file
    d. License and info cells should be merged in table view if possible, and hierarchically ordered in json/data formats 
    e. Example: `0BSD | unencumbered https://spdx.org/licenses/0BSD.html | pacakge.json > foo`
    e. Example: `{ "license": "0BSD", "license_classification_google": "unencumbered",  license_link: "https://spdx.org/licenses/0BSD.html", "packages": ["pacakge.json/foo"] }`
3. Table for usecase#2:
    a. Same as previous except "packages" is now "paths"
    b. since this could be potentially too verbose (e.g when every source code header contains a snippet of the license), we should group related paths that share license, to simplify the view. This can be left for a future enhancement.
4. User can tune the "risky license" threshold by classification: `--license-threshold restricted`
5. User can configure a deny-list of risky licenses which overrides `--license-threshold`: `--license-deny GPL1.0, GPL2.0`

**Comment by owenrumney** [comment]
@knqyf263 - trying to piece together how to get the license information and @itaysk tells me its there already for many, just needs to be surfaced.

I can see `go-dep-parser` provides the license for packages that are stored in the project fs, like `node_modules` but for things like python where it goes to one of a number of possible `site-packages` locations and `go.mod` where it will be in `$GOPATH/pkg/<package>` that doesn't seem have been implemented

Am I missing a way that already exists for more than `node_modules`? 

Itay mentioned `sbom` has licenses?

The other detail I'm looking into, if the user runs `trivy fs --security-checks license .` then we either need to transparently run the `languages` analyser or refactor a bit to allow the license scanner to drive the `go-dep-parser` library for license detection.


TLDR;

1. Is there a way to get license names for packages other than those in `node_modules`
2. Do you agree there will be a need to refactor a bit to suppport `--security-checks license` and still detect packages

**Comment by knqyf263** [comment]
I mean the blue check can be done next. We should focus on the red checks.
![image](https://user-images.githubusercontent.com/2253692/174472944-11f57d15-802e-4b62-aa31-3ca9e4b51e7d.png)

**Comment by itaysk** [comment]
just to clarify the clarification 😆 "should be done next" means deferred for later.

**Comment by itaysk** [comment]
summarizing a convo with @knqyf263 :
the `--license-risk-threshold` flag isn't the best UX (the numerical value isn't self-explanatory) and also not consistent with the rest of Trivy, which is using `--severity`
Suggestion here is to map license classifications into Trivy severities (and use just the severity flag). We need to also show the mapping in the table (both severity and classification).
Proposed mapping:

- Critical
    - non-standard
    - non-commercial (Google's forbidden)
- High
    - restricted
    - restricted_if_statically_linked
- Medium 
    - reciprocal
- Low
    - notice
    - permissive
    - unencumbered
